core/common/src/main/java/alluxio/util/io/PathUtils.java

```diff
@@ -283,17 +291,32 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
    * @throws InvalidPathException when the path or prefix is invalid
    */
   public static boolean hasPrefix(String path, String prefix) throws InvalidPathException {
-    String[] pathComponents = getPathComponents(path);
-    String[] prefixComponents = getPathComponents(prefix);
-    if (pathComponents.length < prefixComponents.length) {
-      return false;
+    if (prefix.isEmpty() || path.isEmpty()) {
+      throw new InvalidPathException(ExceptionMessage.PATH_INVALID.getMessage(""));
     }
-    for (int i = 0; i < prefixComponents.length; i++) {
-      if (!pathComponents[i].equals(prefixComponents[i])) {
-        return false;
-      }
+
+    if (prefix.equals("/")) {
+      return true;
+    }
+
+    if (prefix.endsWith("/") && !path.endsWith("/")) {
+      path = normalizePath(path, "/");
     }
-    return true;
+
+    // Firstly, we check if the `path` starts with `prefix`,
+    // if so, then if `path` and `prefix` is equal, or the prefix endsWith `/` and
+    // the char at the index, which the prefix lastly match, is also a `/`, or the
+    // char at the index right after the `prefix` indicates the end of the last
+    // directory(which is a `/`)
+    // can prove that `prefix` is the prefix of `path`.
+    // The second "or" condition at the second "and" condition is to exclude the
+    // case where `path` starts with the `prefix`, but the last matched position
+    // is actually not the end of a directory(e.g. prefix: '/a/b/c', path: '/a/b/ccc').
+    return path.startsWith(prefix)
+        &&
+        (path.length() == prefix.length()
+            || (prefix.endsWith("/") && path.charAt(prefix.length() - 1) == '/')
+            || path.charAt(prefix.length()) == '/');
```

Comment:
Your comments and the expression are getting too complex. I suggest you break each sub-expression into a boolean variable, and comment on the variable. For example:

```java
// if it starts with hello
boolean isStartsWithHello = str.startsWith("Hello,");
// and ends with world
boolean isEndsWithWorld = str.endsWith("world!");
// then it is probably "Hello, world!"
return isStartsWithHello && isEndsWithWorld;
```

Reply:
Got it. I seperated these conditions and comments.

Reply:
Let's keep using chained booleans because you can get better performance by shorthand.